### ISSUE TITLE
Hyper-parameter search in TensorFlow

### ISSUE THREAD

**Issue by Palang2014** [tensorflow/tensorflow#7868]
Hi all,

Thanks for the great work on TensorFlow.

Is anybody aware of any available TF feature for automatic hyper-parameter search? E.g., using Grid Search, or Random Search or Bayesian Optimization?

Thanks!
Hamid

**Comment by terrytangyuan** [comment]
I think this is something being worked on inside tf.contrib.learn together with Experiment. I don't know the detailed plan though, e.g. whether there will be Bayesian optimized param search method. @martinwicke might know more details.

**Comment by yupbank** [comment]
Hey, is anyone working on this?  I just run into it, cause i need some of this, if no one is working on it, i can help to make a pr :)

**Comment by martinwicke** [comment]
@xiejw FYI.

Let us open-source some more tooling, but then it would be great if someone were to write some Tuning strategies. Or maybe an integration for something like spearmint.

**Comment by martinwicke** [comment]
Take a look at `learn_runner.tune`: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/learn/python/learn/learn_runner.py#L115

In theory, all that needs to be done is to implement one such Tuner object which does (say) grid search, or something more clever. 

I apologize -- the documentation on this is extremely sparse, but please ask questions if the intended use is unclear from the code and comments.

**Comment by kris-singh** [comment]
@martinwicke i also need the same. I will have a shot at this if its okay.

**Comment by martinwicke** [comment]
Go for it. Let me know if you have any questions.

**Comment by kris-singh** [comment]
Can you clarify the create_tuner(study_configuration, objective_key) function. Does it create tuner object? What are the study_configuration, objective_key. I saw the Tuner class which is just an abstract class with a default constructor so where are these parameters used.

**Comment by martinwicke** [comment]
Ah, that's just an example, you don't have to worry about it. The Tuner class is only used in `learn_runner.tune` and you can look at the tiny bit of code that uses it. Basically, you have to implement `Tuner.next_trial()` which asks the tuner to start a new "trial" (an experiment run on specific hyperparameters), and `Tuner.run_experiment`, which takes an experiment_fn with the signature documented in `learn_runner.tune`, and runs the experiments with the hyperparameters for the current trial.

We will add an abstract class to codify the requirements, so far it's all duck typing, yey.

**Comment by Scitator** [comment]
@martinwicke Very cool feature, I am looking forward to adapt it for hyperopt or something like that!
But for now, I stuck with very simple question: is there any simple way to give it ability to run several experiments in parallel (independently)? Or I **must** specify [tf.server](https://www.tensorflow.org/api_docs/python/tf/train/Server#class_tftrainserver) for that? Maybe you can help with some advice or best practices, guide?

**Comment by martinwicke** [comment]
@xiejw for additional comments. A hyperopt tuner would be very cool.

You need `tf.Server` etc. only when you want each experiment to run on several machines. If you have one machine per experiment, you can distribute the python code and run it on each machine independently. You just have to make sure that your Tuner distributes different hyperparameter values to different workers.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by StillerPatrick** [comment]
https://github.com/dandelionmane/tf-dev-summit-tensorboard-tutorial/blob/master/mnist.py

This guy from google implemented a small hyperparameter search. Maybe it helps.

**Comment by makoeppel** [comment]
For one of my projects I created a gridsearch similar to the one from  scikit-learn (http://scikit-learn.org/stable/modules/grid_search.html).

You can have a look at https://github.com/makoeppel/hyperParameterSearchTF/tree/master
--> Please take in mind that for now the classes are not filled with a specific model now and the repo is still under contraction. I tested everything with my own models so you just need to fill the empty classes with your specific model.

**The program works like this:**

The idea here is that you configure your search in the main.py. There you load your someClf.clfHandler and your evaluator.evaluatorHandler. With some default values. In the example code it is: 

`someClf = someClf.clfHandler(outputDir, outputName, layers=[5, 5, 1], feature_func=difference,
                                         weight_func=summed_weight, cost_func=weighted_least_square,
                                         optimizer=tf.train.AdamOptimizer(0.01), activation=tf.nn.tanh,
                                         use_bias=False, kernel_initializer=tf.random_normal_initializer(),
                                         steps=20, max_iters=2000, n_samples=1000, max_samples=3000)`

and 

`clfEvaluator = evaluator.evalHandler(inputDir=outputDir,
                                         inputName=outputName,
                                         feature_func=difference,
                                         weight_func=summed_weight
                                         )`

After this you pass everything to the multimodelsearch.MultiModelSearch. This class will generate for every parameter in your parameter dict clfHandler and evalHandler classes with the default values and will replace the default ones  with the ones from the dict.

The function MultiModelSearch.fit_and_eval() is using the multiprocessing library and will train and evaluate all generated models and after this it will return a dict with the results. 

At the end I used this file (http://scikit-learn.org/stable/auto_examples/model_selection/plot_multi_metric_evaluation.html#sphx-glr-auto-examples-model-selection-plot-multi-metric-evaluation-py) for plotting the results. 

If you need more informations please ask me. I am open for contribution.

**Comment by makoeppel** [comment]
I used the mnist.py example from https://www.tensorflow.org/get_started/mnist/beginners and put that into my code now (https://github.com/makoeppel/hyperParameterSearchTF/tree/master).

**Comment by makoeppel** [comment]
I now created this tuner class (https://github.com/makoeppel/tuner_object_for_learn_runner) and pass it to the `learn_runner.tune(experiment_fn=experiment_fn,` `tuner=tuner)` function. I used an example from https://gist.github.com/peterroelants/9956ec93a07ca4e9ba5bc415b014bcca. 

The tuner class can be called with `default_param, hparams, model_dir.` When you call `next_trial()` it set everything to the `default_param` and goes to the next `hparam` in `hparams` and replaces the `default_param` with this one. When you call `run_experiment()` it runs 
`learn_runner.run(experiment_fn=experiment_fn, 
            run_config=self.run_config,
            schedule="train_and_evaluate", 
            hparams=hparams)` with the hparams after the current next_trial() call. So it is a really basic gridsearch.

**Comment by Hvass-Labs** [comment]
I almost never advertise my tutorials, but I came across this thread and saw it was still active, so you might like to know about my recent tutorial on hyper-parameter optimization in TensorFlow using Keras and scikit-optimize which has various forms of Bayesian optimization methods including Gaussian Processes and Random Forests:

https://github.com/Hvass-Labs/TensorFlow-Tutorials/blob/master/19_Hyper-Parameters.ipynb

https://www.youtube.com/watch?v=oaxf3rk0KGM

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by asprenger** [comment]
I need hyperparameter search for one of my projects and could spend some time on this issue. I did some prototyping for a single machine with multiple GPUs where a model instance is naively assigned to each GPU: [https://github.com/asprenger/tf-param-search/blob/master/param_search.py](https://github.com/asprenger/tf-param-search/blob/master/param_search.py). The parameter sets are currently generated using classes from Sklearn model selection. In the end I would like to be able to distribute parameter search over multiple machines. Maybe someone could provide some feedback.

**Comment by badlogicmanpreet** [comment]
Hi @martinwicke is this issue being worked on or can I give it a try? Thanks

**Comment by martinwicke** [comment]
There are several efforts around hyperparameter search for Keras already. Did you have a specific project in mind?

**Comment by badlogicmanpreet** [comment]
Thanks @martinwicke , I was thinking of Bayesian optimization based method. As you rightly said it must be already part of the efforts. Wanted to mainly start as a beginner on the contribution, since I use these techniques I thought contributing here will be a good start.
Will explore other issues on contribute label to start with.

**Comment by martinwicke** [comment]
Yeah, this is a pretty complicated issue to start with. Thanks for getting
involved!

**Comment by dynamicwebpaige** [comment]
Going to close this issue, as [Keras Tuner](https://github.com/keras-team/keras-tuner) will support RandomSearch, Hyperband, and Bayesian optimization. Thank you for the request, and be on the lookout for KT's 1.0 release!

**Comment by sumanthratna** [comment]
Is there a recommended way to approach Bayesian optimization without Keras? I'm currently using the TensorBoard hparams plugin but I can only do grid/random search with that.
